### PR TITLE
DAOS-4441 Placement: Modify Placement for OS Reintegration

### PR DESCRIPTION
**Pull request by PetFet**
Modified the placement API functionality so that now when
a reintegration operation is ongoing it will temporarily extend
the layout to include the reintegrating targets.

Signed-off-by: Peter Fetros <peter.fetros@intel.com>

### CODE REVIEW

**Comment by PetFet** [comment]
https://github.com/daos-stack/daos/pull/2284 Needs to be landed first if possible.

File: src/placement/jump_map.c [hunk 1]
```diff
@@ -31,6 +31,14 @@
 #include <daos/pool_map.h>
 #include <isa-l.h>
 
+enum PL_OP_TYPE {
+	PL_PLACE,
+	PL_PLACE_EXTENDED,
+	PL_REBUILD,
+	PL_REINT,
+	PL_ADD,
+};
```
Comment (NiuYawei) [inline src/placement/jump_map.c:40]: It's better to put some comments to explain these ops, that'll be helpful for understanding the code.

**Comment by PetFet** [comment]
> what is the difference between PL_PLACE_EXTENDED and PL_ADD?

Extend is used for calculating the placement during a reintegration or addition operation. We return an "extended" layout which contains the reintegrating nodes so that we can write new data to them during reintegration.


Add will be used if needed for addition. Currently it doesn't have a use, but will as soon as I implement PL_MAP_ADD or related functionality.

File: src/placement/jump_map.c [hunk 2]
```diff
@@ -600,17 +637,30 @@ get_object_layout(struct pl_jump_map *jmap, struct pl_obj_layout *layout,
 			/** If target is failed queue it for remap*/
 			if (pool_target_unavail(target, for_reint)) {
```
[the remark this answers is not in the document]
Comment (wangdi1) [reply src/placement/jump_map.c:646]: I see. thanks for clarification.

**Comment by wangdi1** [comment]
> > what is the difference between PL_PLACE_EXTENDED and PL_ADD?
> 
> Extend is used for calculating the placement during a reintegration or addition operation. We return an "extended" layout which contains the reintegrating nodes so that we can write new data to them during reintegration.
> 
> Add will be used if needed for addition. Currently it doesn't have a use, but will as soon as I implement PL_MAP_ADD or related functionality.

Hmm, what use cases of addition do you mean here?  It looks like to me reintegration and addition are same for each individual object with regard to layout anyway?  do I miss sth?

**Comment by PetFet** [comment]
> > > what is the difference between PL_PLACE_EXTENDED and PL_ADD?
>> 
>> Extend is used for calculating the placement during a reintegration or addition operation. We return an "extended" layout which contains the reintegrating nodes so that we can write new data to them during reintegration.
>> 
>> Add will be used if needed for addition. Currently it doesn't have a use, but will as soon as I implement PL_MAP_ADD or related functionality.

>Hmm, what use cases of addition do you mean here?  It looks like to me reintegration and addition are same for each individual object with regard to layout anyway?  do I miss sth?

When calculating addition as part of the re-balance/migration process we will need to first calculate the original location of the shards as well as where they moved to. 

finding the fallback target in reintegration would be a different process then finding where the previous target was in addition.

This will probably require calculating layout using a slightly different method. with that in mind I just added the "ADD" opcode because I thought I would need it when implementing Addition

**Comment by PetFet** [comment]
I believe the test failures are related to this https://jira.hpdd.intel.com/browse/DAOS-5050
among other intermittent tests.

**Comment by jolivier23** [comment]
Are you sure you didn't introduce a failure here?

Test / Functional_Hardware_Medium / IO40: Record count after punch/enumeration – DAOS IO tests
<1s
Stacktrace
ASSERT: obj_shard->do_shard == shard
src/object/cli_obj.c:143: error: Failure!

I see you code does actually set this field

**Comment by PetFet** [comment]
> Are you sure you didn't introduce a failure here?
> 
> Test / Functional_Hardware_Medium / IO40: Record count after punch/enumeration – DAOS IO tests
> <1s
> Stacktrace
> ASSERT: obj_shard->do_shard == shard
> src/object/cli_obj.c:143: error: Failure!
> 
> I see you code does actually set this field

I was able to run this successfully on boro but I'll dig into the logs some more and get back to you.

**Comment by PetFet** [comment]
Copying the fixes commit message here. Was harder to track down because the tests that were reported as failed were not actually the ones causing the problem.

Fixed several edge cases related to new extended layout sometimes
returned during the test.

 - There was a problem with the new extended layout being returned
   during the IO test. This was because the IO test was triggering
   reintegration by making a call to pool_add_target which had not triggered
   reintegration in the past but was now causing a race between the rebuild
   process and the IO test *among other checks inconsistent with the new layout ordering.

- It was also possible that while choosing the leader during placement,
  the newly extended shard would be chosen as the leader when it
  shouldn't have been. Now any shard which is not yet reintegrated will
  be skipped as the leader.

**Comment by PetFet** [comment]
OID Test Failure and Some Intermittent Rebuild Failures. These one's I do think are intermittent failures that are affecting everyone.

**Comment by PetFet** [comment]
Previously failed on rebuild Test 22, This can be found here https://jira.hpdd.intel.com/browse/DAOS-3952

File: src/object/cli_obj.c [hunk 3]
```diff
@@ -139,9 +139,6 @@ obj_shard_open(struct dc_object *obj, unsigned int shard, unsigned int map_ver,
 		D_GOTO(unlock, rc = -DER_NONEXIST);
 	}
 
-	/* XXX could be otherwise for some object classes? */
-	D_ASSERT(obj_shard->do_shard == shard);
```
Comment (NiuYawei) [inline src/object/cli_obj.c:143]: In what use case they are not equal?

File: src/placement/tests/place_obj_common.c [hunk 4]
```diff
@@ -365,10 +406,7 @@ plt_reint_tgts_get(uuid_t pl_uuid, daos_obj_id_t oid, uint32_t *failed_tgts,
 	D_ASSERT(rc >= 0);
 	*spare_cnt = rc;
 
-	D_PRINT("reint_cnt %d for version %d -\n", *spare_cnt, *po_ver);
 	for (i = 0; i < *spare_cnt; i++)
-		D_PRINT("shard %d, spare target rank %d\n",
-			shard_ids[i], spare_tgt_ranks[i]);
```
Comment (NiuYawei) [inline src/placement/tests/place_obj_common.c:410]: that'll leave an empty for loop?

File: src/object/cli_obj.c (hunk 3 above)
Comment (PetFet) [reply src/object/cli_obj.c:143]: In this patches scenario.

When performing online features (reintegration/drain/etc.) we need to return an extended layout containing the shards we're moving data to, that way when we perform concurrent I/O the new writes also go to the targets we're reintegrating.

In this case we need a way keep track of which shards are in the extended layout by giving them the same shard number.

For example
```
[T1 T2 T3] <- original layout.
[S1 S2 S3] <- shard ordering

[T4 T2 T3] <- new layout after failure and rebuild.
[S1 S2 S3]

[T4 T2 T3 T1] <- for online reintegration, we need to write to everything because of concurrent I/O.
[S1 S2 S3 S1] <- Reintegrating target being reintegrated for shard 1

[T1 T2 T3] <- reintegration complete.
[S1 S2 S3]
```

File: src/placement/tests/place_obj_common.c (hunk 4 above)
Comment (PetFet) [reply src/placement/tests/place_obj_common.c:410]: That was a mistake, thanks for catching it!

File: src/placement/pl_map_common.c [hunk 5]
```diff
@@ -399,4 +399,94 @@ determine_valid_spares(struct pool_target *spare_tgt, struct daos_obj_md *md,
...
+		current = current->next;
+	}
+
+
+	new_group_size = layout->ol_grp_size + max_fail_grp;
+	D_ALLOC_ARRAY(new_shards, new_group_size * layout->ol_grp_nr);
+	if (new_shards == NULL)
+		return -DER_NOMEM;
+
+	while (k < layout->ol_nr) {
```
Comment (gnailzenh) [inline src/placement/pl_map_common.c:455]: easier to understand in this way? " for (i = k = 0; k < layout->ol_nr;)"
if we need to set initial value of loop, I think it's cleaner to just use "for" statement

**Comment by PetFet** [comment]
Currently failing Rebuild Test #3.
The client log reports this error:

`06/15-22:21:58.06 wolf-52 DAOS[32257/32257] object INFO src/object/obj_verify.c:539 dc_obj_verify_cmp() 1152921590623633466.31 (reps 3, inconsistent) shard 0 has rec type 0, but shard 1 has rec type 4.
06/15-22:21:58.06 wolf-52 DAOS[32257/32257] object ERR  src/object/obj_verify.c:710 dc_obj_verify_rdg() Failed to verify cmp: DER_MISMATCH(-1031)`

Di is working on a patch for this issue (DAOS-4623)
